imblearn/scaling/css.py

```diff
@@ -0,0 +1,249 @@
+                            CSS_MODE, self.mode))
+
+        if self.target not in CSS_TARGET:
+            raise ValueError('Unknown kind for CSS target.'
+                             ' Choices are {}. Got \'{}\' instead.'.format(
+                                CSS_TARGET, self.target))
+
+        if self.c < 0 or self.c > 1:
+            raise ValueError('Received scaling factor c={}, which'
+                             ' is outside the allowed range '
+                             '(0-1].'.format(self.c))
+        if self.c is 0:
+            raise ValueError('Received scaling factor c={}, which is'
+                             ' equal to no CSS at.'.format(self.c))
+
+        if self.minority_class_value is not None and \
+                not isinstance(self.minority_class_value, int):
+            raise ValueError('Unallowed target class value \'{}\'.'
+                             ' Valid values include None to automatically'
+                             ' infer the target class or any integer number'
+                             ' corresponding to the value of the label in y')
+
+
+        mcv = self.minority_class_value
+        if mcv is None:
+            # infer minority class value
+            counts = Counter(y)
+            least_common = counts.most_common()[:-1-1:-1]
+            mcv = least_common[0][0]
+
+        # in the following _a is majority, _i is minority
+        if self.target is "majority" or self.target is "both":
+            col_means_a = np.mean(X[(y != mcv)], axis=0)
+            if self.mode is "linear":
+                distances_a = abs(np.subtract(X[y != mcv], col_means_a))
+        if self.target is "minority" or self.target is "both":
+            col_means_i = np.mean(X[(y == mcv)], axis=0)
+            if self.mode is "linear":
+                distances_i = abs(np.subtract(X[y == mcv], col_means_i))
+
+        if self.target is "majority" or self.target is "both":
+            if self.mode is "constant":
+                X_scaled_a = X[y != mcv] * (1 - self.c) + col_means_a * self.c
+            elif self.mode is "linear":
+                scale_factors_mean = (distances_a * self.c)
+                scale_factors_values = (1 - self.c * distances_a)
+
+                X_scaled_a = X[y != mcv] * scale_factors_values + col_means_a * scale_factors_mean
+        if self.target is "minority" or self.target is "both":
+            if self.mode is "constant":
+                X_scaled_i = X[y == mcv] * (1 - self.c) + col_means_i * self.c
+            elif self.mode is "linear":
+                scale_factors_mean = (distances_i * self.c)
+                scale_factors_values = (1 - self.c * distances_i)
+                X_scaled_i = X[y == mcv] * scale_factors_values + col_means_i * scale_factors_mean
+
+        # merge scaled and non scaled stuff
+        if self.target is "majority":
+            X_scaled = np.concatenate([X_scaled_a, X[y == mcv]], axis=0)
```

Comment:
you need to use `safe_indexing` from scikit-learn to be able to use sparse matrices